### ISSUE TITLE
QIR: compilation fails and no .log file is generated

### ISSUE THREAD

**Issue by esthercruz** [microsoft/qsharp-compiler#847]
I am trying to generate QIR (using current version of feature/qir branch) from an empty file foo.qs with the following command from the root directory :

`dotnet run --project src/QsCompiler/CommandLineTool/ build --qir QIR s --input  foo.qs examples/QIR/QirCore.qs examples/QIR/QirTarget.qs --proj BOO`

I am getting the following error: 

```
Error QS7109: 
The compilation step "QIR Generation" loaded from "C:\Users\esthe\qsharp-compiler\src\QsCompiler\CommandLineTool\bin\Debug\netcoreapp3.1\Microsoft.Quantum.QsCompiler.dll" threw an exception.

____________________________________________

Q# compilation failed: 1 error, 1 warning 
1 logged exception
```

An empty .ll file is generated, but no .log file is. Any idea about the reason behind both the exception and the missing .log file?

Using dotnet version 3.1.404

**Comment by samarsha** [comment]
Hi, could you try running the compiler in verbose mode? This should show some more info about the exception:

`dotnet run -- --verbosity detailed --project src/QsCompiler/CommandLineTool/ build --qir QIR s --input foo.qs examples/QIR/QirCore.qs examples/QIR/QirTarget.qs --proj BOO`

**Comment by cgranade** [comment]
@samarsha: I'm not sure if it's helpful or not, but I run into similar errors running on my machine against the `feature/qir` branch as well. Setting `-v diag` gives an exception in each of the C# and QIR code generation steps:

```
System.TypeLoadException: Could not load type 'Microsoft.Quantum.QsCompiler.SyntaxExtensions' from assembly 'Microsoft.Quantum.QsDataStructures, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at Microsoft.Quantum.QsCompiler.CsharpGeneration.CodegenContext.Create(IEnumerable`1 syntaxTree, IDictionary`2 assemblyConstants)  
   at Microsoft.Quantum.QsCompiler.CsharpGeneration.CodegenContext.Create(QsCompilation compilation, IDictionary`2 assemblyConstants) 
in D:\a\1\s\submodules\qsharp-runtime\src\Simulation\CsharpGeneration\Context.fs:line 99
   at Microsoft.Quantum.QsCompiler.CsharpGeneration.Emitter.Microsoft-Quantum-QsCompiler-IRewriteStep-Transformation(QsCompilation compilation, QsCompilation& transformed) in D:\a\1\s\submodules\qsharp-runtime\src\Simulation\CsharpGeneration\RewriteStep.fs:line 51    
   at Microsoft.Quantum.QsCompiler.LoadedStep.Transformation(QsCompilation compilation, QsCompilation& transformed) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Compiler\LoadedStep.cs:line 194
   at Microsoft.Quantum.QsCompiler.CompilationLoader.ExecuteRewriteStep(LoadedStep rewriteStep, QsCompilation compilation, QsCompilation& transformed) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Compiler\CompilationLoader.cs:line 909       


Error QS7109:
The compilation step "CsharpGeneration" loaded from "C:\Users\cgran\.nuget\packages\microsoft.quantum.csharpgeneration\0.15.2101125897\lib\netstandard2.1\Microsoft.Quantum.CsharpGeneration.dll" threw an exception.


System.ArgumentException: invalid loop variable name
   at Microsoft.Quantum.QsCompiler.QIR.QirStatementKindTransformation.<>c__DisplayClass12_0.<OnForStatement>g__ExecuteBody|0(Value loopVariable) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Subtransformations\StatementKindTransformation.cs:line 327
   at Microsoft.Quantum.QsCompiler.QIR.GenerationContext.<>c__DisplayClass82_1.<CreateForLoop>b__4() in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Context.cs:line 1462
   at Microsoft.Quantum.QsCompiler.QIR.GenerationContext.<>c__DisplayClass82_0.<CreateForLoop>g__PopulateLoopBody|1(Action executeBody) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Context.cs:line 1437
   at Microsoft.Quantum.QsCompiler.QIR.GenerationContext.<>c__DisplayClass82_0.<CreateForLoop>b__3() in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Context.cs:line 1462
   at Microsoft.Quantum.QsCompiler.QIR.GenerationContext.ExecuteLoop(BasicBlock continuation, Action loop) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Context.cs:line 1480
   at Microsoft.Quantum.QsCompiler.QIR.GenerationContext.CreateForLoop(Value startValue, Func`2 evaluateCondition, Value increment, Action`1 executeBody) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Context.cs:line 1459        
   at Microsoft.Quantum.QsCompiler.QIR.GenerationContext.IterateThroughRange(Value start, Value step, Value end, Action`1 executeBody) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Context.cs:line 1540
   at Microsoft.Quantum.QsCompiler.QIR.QirStatementKindTransformation.OnForStatement(QsForStatement stm) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Subtransformations\StatementKindTransformation.cs:line 336
   at Microsoft.Quantum.QsCompiler.Transformations.Core.StatementKindTransformationBase.OnStatementKind(QsStatementKind kind) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\StatementTransformation.fs:line 229
   at Microsoft.Quantum.QsCompiler.Transformations.Core.StatementTransformationBase.OnStatement(QsStatement stm) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\StatementTransformation.fs:line 316
   at <StartupCode$Microsoft-Quantum-QsCore>.$StatementTransformation.statements@327.Invoke(QsStatement arg00) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\StatementTransformation.fs:line 327
   at Microsoft.FSharp.Collections.Internal.IEnumerator.map@75.DoMoveNext(b& curr) in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 78
   at Microsoft.FSharp.Collections.Internal.IEnumerator.MapEnumerator`1.System-Collections-IEnumerator-MoveNext() in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 64
   at Microsoft.FSharp.Collections.SeqModule.Iterate[T](FSharpFunc`2 action, IEnumerable`1 source) in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 494
   at Microsoft.Quantum.QsCompiler.Transformations.Core.StatementTransformationBase.OnScope(QsScope scope) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\StatementTransformation.fs:line 332
   at Microsoft.Quantum.QsCompiler.QIR.QirNamespaceTransformation.OnProvidedImplementation(QsTuple`1 argTuple, QsScope body) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Subtransformations\NamespaceTransformation.cs:line 73    
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnSpecializationImplementation(SpecializationImplementation implementation) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 185
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnSpecializationKind(QsSpecialization spec) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 200
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnBodySpecialization(QsSpecialization spec) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 208
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnSpecializationDeclaration(QsSpecialization spec) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 223     
   at Microsoft.Quantum.QsCompiler.QIR.QirNamespaceTransformation.OnSpecializationDeclaration(QsSpecialization spec) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Subtransformations\NamespaceTransformation.cs:line 81
   at <StartupCode$Microsoft-Quantum-QsCore>.$NamespaceTransformation.specializations@243.Invoke(QsSpecialization arg00) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 243
   at Microsoft.FSharp.Collections.Internal.IEnumerator.map@75.DoMoveNext(b& curr) in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 78
   at Microsoft.FSharp.Collections.Internal.IEnumerator.MapEnumerator`1.System-Collections-IEnumerator-MoveNext() in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 64
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Collections.Immutable.ImmutableArray.CreateRange[T](IEnumerable`1 items)
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnCallableKind(QsCallable c) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 240
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnOperation(QsCallable c) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 253
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnCallableDeclaration(QsCallable c) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 266
   at Microsoft.Quantum.QsCompiler.QIR.QirNamespaceTransformation.OnCallableDeclaration(QsCallable c) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Subtransformations\NamespaceTransformation.cs:line 89
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnNamespaceElement(QsNamespaceElement element) in 
C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 294
   at <StartupCode$Microsoft-Quantum-QsCore>.$NamespaceTransformation.elements@311.Invoke(QsNamespaceElement arg00) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 311
   at Microsoft.FSharp.Collections.Internal.IEnumerator.map@75.DoMoveNext(b& curr) in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 78
   at Microsoft.FSharp.Collections.Internal.IEnumerator.MapEnumerator`1.System-Collections-IEnumerator-MoveNext() in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 64
   at Microsoft.FSharp.Collections.SeqModule.Iterate[T](FSharpFunc`2 action, IEnumerable`1 source) in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 494
   at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnNamespace(QsNamespace ns) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 316
   at Microsoft.Quantum.QsCompiler.QIR.Generator.Apply() in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\QirGeneration\Generator.cs:line 69
   at Microsoft.Quantum.QsCompiler.BuiltInRewriteSteps.QirGeneration.Transformation(QsCompilation compilation, QsCompilation& transformed) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Compiler\RewriteSteps\QirGeneration.cs:line 72
   at Microsoft.Quantum.QsCompiler.LoadedStep.Transformation(QsCompilation compilation, QsCompilation& transformed) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Compiler\LoadedStep.cs:line 194
   at Microsoft.Quantum.QsCompiler.CompilationLoader.ExecuteRewriteStep(LoadedStep rewriteStep, QsCompilation compilation, QsCompilation& transformed) in C:\Users\cgran\source\repos\microsoft\qsharp-compiler\src\QsCompiler\Compiler\CompilationLoader.cs:line 909
```

**Comment by Co0olboi** [comment]
- [ ]

**Comment by bettinaheim** [comment]
@esthercruz @cgranade That the C# compiler fails when QIR emission is enable is currently a known issue. 
"Could not load type 'Microsoft.Quantum.QsCompiler.SyntaxExtensions'" is a mismatch between the runtime version and the compiler version that you are using, and "System.ArgumentException: invalid loop variable name" is the result of an invalid syntax tree (possibly a subsequent error from the first one). 
We are working hard on getting this to main, after which things should become a lot more easy to use, and artifacts from a runtime/compiler mismatch shouldn't be a problem anymore. We should get there within the next week or so.
@esthercruz If you don't want to wait until then I am happy to take a look if you post the output of the verbose mode.

**Comment by bettinaheim** [comment]
@cgranade Sorry, the "invalid loop variable name" is not a subsequent error, but an issue with discarded loop variables for iterations over ranges. It will be fixed with the next PR (it's currently on beheim/entryPoints).

**Comment by cgranade** [comment]
@bettinaheim: Can confirm, running against `beheim/entryPoints` fixed that error. Thank you for the help on that!

**Comment by esthercruz** [comment]
Thanks all for your help and sorry for my late response! 

@bettinaheim @samarsha I've run the command in verbosity mode. There's a first phase that shows no error and end with "Build completed". After that the same error as I got at the beginning still appears, with no more detail. Please find the complete output here: https://gist.github.com/esthercruz/4e79256340f7014208ca1015db78cd8b

I don't if this helps to understand the problem since no error seems to appear, so if we don't get any solution I can wait until the mismatch runtime/compiler is fixed. Thanks a lot! :)

**Comment by crazy4pi314** [comment]
I don't know if it is helpful here, but I have been able to get a teleport Q# program to compile with the version of the compiler that is on `feature/qir`  and put the sample program and a readme with the steps to repro here: https://github.com/esthercruz/qosf_mentorship_project/tree/main/teleport-qir 

**tl;dr** both building the compiler works (but fails on other tools that are built in the same process after the compiler itself), and then the compilation of the sample teleport program to QIR fails to compile the C# output but only after it has successfully emitted the QIR file. So I can go from Q#->QIR successfully but most processes don't cleanly exit (and I think both are know to the dev team).

**Comment by bettinaheim** [comment]
@esthercruz Support for QIR emission is now on main. Please see [this readme](https://github.com/microsoft/qsharp-compiler/tree/main/src/QsCompiler/QirGeneration#qir-emission---preview-feature) for more information on how to use it. Please let me know if that resolves your issue, or if you are still running into the same (or new) problems.

**Comment by bettinaheim** [comment]
@esthercruz The issue as it is stated is obsolete with the changes we did to pull the QIR emission into main. Please feel free to reopen with an update or open a new issue if you still run into problems.
